### ISSUE TITLE
Make python binding generator more generic

### ISSUE THREAD

**Issue by dov** [Itseez/opencv#6599]
The current python binding generator gen2.py makes a great job for creating python bindings of the core opencv library. Unfortunately it fails to help in the creation of third party libraries that links to opencv. An example of a such a third party library is http://opencvblobslib.github.io/opencvblobslib/ .

There are two issues involved:

1. The python binding generator does not create .h files that a third party library can use. E.g. lets say that we have a library Foo wrapped into the python module PyFoo. We also have a library Bar with a python module PyBar(). For PyBar to access PyFoo, we need an h-file similar to the below. This allows a third party both to receive and decompose python objects as well as constructing new ones as return values.

```
#ifndef PYFOO_H
#define PYFOO_H

#include <Python.h>
#include <Foo.h>

PyObject *PyFoo_FromFoo(Foo foo);

typedef struct {
    PyObject_HEAD;
    Foo *foo;
} PyFoo;

#define PyFoo_Check(op) \
    PyObject_TypeCheck(op, &PyFooType)

extern PyTypeObject PyFooType;
#endif
```

(This corresponds to the `pyopencv_from()` and `pyopencv_to()` functions that need to be made visible in the h-files. I'm not sure that I like the idea of having public API that rely on template instantiated functions. Why not just create a mangled name that contains the type of the from and to?)

2. Make `gen2.py` able to generate modules with a user chosen module name. This involved adding command line parameters for lots of hardwired module names in gen2.

**Comment by pixelou** [comment]
Hi @dov,
I had a relatively similar use case for a project and tried to engage [a discussion on that topic](http://answers.opencv.org/question/92353/how-to-generate-python-bindings-for-opencv-based-projects/). If that can help you: I have also extracted [a slightly more generic version of gen2](https://github.com/pixelou/CPPyModule), but this probably needs more care to be fully reliable. In absence of reaction, I believed it was not that important for the community.

**Comment by mshabunin** [comment]
In case when you need interoperability between native libraries hidden behind python or java I'd recommend to wrap your library into personal OpenCV module and build them together. Different compilers, flags or standard libraries can cause more problems than benefits.

For example, you have a library Foo which uses OpenCV and you want to call it from python, then:
- add minimal module stuff: CMakeLists.txt, directory structure (`Foo/ocv/modules/foo`)
- add wrapper C++ headers with `CV_EXPORTS_W` directives to `<module>/include/opencv2` folder
- add manual converters to the `<module>/misc/python` if needed
- build OpenCV with your module and python support: `cmake -DOPENCV_EXTRA_MODULES_PATH=<some-path>/Foo/ocv ../opencv && make`

BTW, related update to the python wrapping process has been introduced recently: #7056  - now each module can provide manual-written conversion routines in `<module>/misc/python` folder.

I did not try the described scheme and could miss something, but developing standalone automatic universal C++-to-python wrapping mechanism from hdr_parser.py and gen2.py looks like quite hard task.

**Comment by pixelou** [comment]
I agree that it would be a hard thing to maintain one as a standalone generic project. The issue is for a slightly more generic wrapping tool for OpenCV-style third-party projects though. Nevertheless, I'd argue that this task is rather independant from the core OpenCV project unless you guys want to help growing the module ecosystem beyond the main project. This is not always beneficial because support often becomes unreliable.

For what it's worth, cython's wiki points to several tools that have been developped to address automatic bindings generation:
- https://github.com/cython/cython/wiki/WrappingCorCpp
- https://github.com/cython/cython/wiki/AutoPxd
